simperby/src/node/stf.rs

```diff
@@ -0,0 +1,288 @@
+            index,
+            message: "validator does not exist".to_string(),
+        })?;
+    delegation_state
+        .original_validator_set
+        .remove(validator_index);
+    for validator in delegation_state.original_validator_set.iter_mut() {
+        if let Some((_, current_delegatee)) = validator.2 {
+            if current_delegatee == validator_index {
+                validator.2 = None;
+            }
+        }
+    }
+
+    storage.update_delegation_state(delegation_state).await?;
+    Ok(())
+}
+
+async fn execute_delegate(
+    storage: &mut StateStorage,
+    last_header: &BlockHeader,
+    index: usize,
+    delegator: PublicKey,
+    delegatee: PublicKey,
+    target_height: BlockHeight,
+    commitment: TypedSignature<(PublicKey, BlockHeight)>,
+) -> Result<(), ExecutionError> {
+    if target_height != last_header.height + 1 {
+        return Err(ExecutionError::TransactionFailure {
+            index,
+            message: format!(
+                "target height does not match: expected {}, got {}",
+                target_height,
+                last_header.height + 1
+            ),
+        });
+    }
+    if delegatee == delegator {
+        return Err(ExecutionError::TransactionFailure {
+            index,
+            message: "delegator and delegatee are the same".to_string(),
+        });
+    }
+    commitment
+        .verify(&(delegatee.clone(), target_height), &delegator)
+        .map_err(|e| ExecutionError::TransactionFailure {
+            index,
+            message: format!("delegation signature verification failed: {}", e),
+        })?;
+
+    let mut delegation_state = storage.get_delegation_state().await?;
+    let delegator_index = delegation_state
+        .original_validator_set
+        .iter()
+        .position(|(x, _, _)| *x == delegator)
+        .ok_or_else(|| ExecutionError::TransactionFailure {
+            index,
+            message: "delegator does not exist".to_string(),
+        })?;
+    let delegatee_index = delegation_state
+        .original_validator_set
+        .iter()
+        .position(|(x, _, _)| *x == delegatee)
+        .ok_or_else(|| ExecutionError::TransactionFailure {
+            index,
+            message: "delegatee does not exist".to_string(),
+        })?;
+
+    // update current_delegatee for all its delegators, if exist. (i.e., re-delegate)
+    for validator in delegation_state.original_validator_set.iter_mut() {
+        if let Some((_, current_delegatee)) = validator.2.as_mut() {
+            if *current_delegatee == delegator_index {
+                *current_delegatee = delegatee_index;
+            }
+        }
+    }
+
+    if let Some(x) = delegation_state.original_validator_set[delegator_index].2 {
+        return Err(ExecutionError::TransactionFailure {
+            index,
+            message: format!(
+                "delegator already has delegateed ({:?}); please undelegate first",
```

Comment:
Why do we force the undelegation before the delegation?

Reply:
I have the similar question.
Any concern with replacing a delegatee within a single block?

Reply:
@lemonholic @byeongjee @junha1  Is it okay in notion to delegate to another delegatee while delegator has already delegated?

Reply:
@lemonholic for explicity
@byeongjee we can do it in a single block (using multiple txes)

Reply:
ok